<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-Sample-Data" data-toc-modified-id="Import-Sample-Data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import Sample Data</a></span><ul class="toc-item"><li><span><a href="#Housing-Price" data-toc-modified-id="Housing-Price-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Housing Price</a></span></li><li><span><a href="#Tax-Data" data-toc-modified-id="Tax-Data-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Tax Data</a></span></li></ul></li><li><span><a href="#join-Data-Frame-Method" data-toc-modified-id="join-Data-Frame-Method-2"><span class="toc-item-num">2&nbsp;&nbsp;</span><code>join</code> Data Frame Method</a></span></li><li><span><a href="#concat-Pandas-Function" data-toc-modified-id="concat-Pandas-Function-3"><span class="toc-item-num">3&nbsp;&nbsp;</span><code>concat</code> Pandas Function</a></span></li><li><span><a href="#merge-Data-Frame-Method" data-toc-modified-id="merge-Data-Frame-Method-4"><span class="toc-item-num">4&nbsp;&nbsp;</span><code>merge</code> Data Frame Method</a></span></li></ul></div>

In [13]:
import pandas as pd

## Import Sample Data

### Housing Price
The Housing Price data by zip from Zillow was downloaded from https://www.zillow.com/research/data/ .  


In [14]:
# house_df = pd.read_csv('../data/Zip_zhvi_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv')
house_df = pd.read_csv('https://raw.githubusercontent.com/emsanqui/Python_Training/master/data/Zip_zhvi_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv')
house_df.tail()

RegionID  SizeRank  RegionName RegionType StateName State         City   
27010     78282     39809       46799        zip        IN    IN   Zanesville  \
27011     80861     39809       52163        zip        IA    IA     Protivin   
27012     69074     39809       26576        zip        WV    WV   Farmington   
27013     64515     39809       16245        zip        PA    PA    Oak Ridge   
27014     80190     39809       50160        zip        IA    IA  Martensdale   

                                Metro        CountyName  2000-01-31  ...   
27010                    Bluffton, IN      Wells County         NaN  ...  \
27011                             NaN     Howard County         NaN  ...   
27012                    Fairmont, WV     Marion County         NaN  ...   
27013                  Pittsburgh, PA  Armstrong County         NaN  ...   
27014  Des Moines-West Des Moines, IA     Warren County         NaN  ...   

          2022-06-30     2022-07-31     2022-08-31     2022-09-30   
27010  172931.113553  173726.818013  173951.262842  173561.093030  \
27011  100986.105306  102843.677816  104486.379416  105810.880094   
27012   98430.653830   99565.281599  100539.187287  100590.832913   
27013   79001.623917   79797.115905   80327.757426   80096.167777   
27014  182921.063195  184139.412398  184688.403061  184729.848799   

          2022-10-31     2022-11-30     2022-12-31     2023-01-31   
27010  173267.350240  173316.329830  173324.855024  173343.968822  \
27011  106454.239900  106897.177835  107196.786820  107037.013712   
27012  100489.571763  100123.791299  100196.541945  100556.335407   
27013   79256.847170   78066.159714   77466.901397   77711.217346   
27014  185179.263239  185869.441655  186269.605436  186267.600047   

          2023-02-28     2023-03-31  
27010  173739.502719  174765.029798  
27011  106366.668071  105680.453415  
27012  101320.684556  102093.120933  
27013   77581.056657   76642.895385  
27014  186353.463783  187071.766317  

[5 rows x 288 columns]

Since the housing data has 288 columns, we will select a few columns to work with.

In [15]:
house_df.columns

Index(['RegionID', 'SizeRank', 'RegionName', 'RegionType', 'StateName',
       'State', 'City', 'Metro', 'CountyName', '2000-01-31',
       ...
       '2022-06-30', '2022-07-31', '2022-08-31', '2022-09-30', '2022-10-31',
       '2022-11-30', '2022-12-31', '2023-01-31', '2023-02-28', '2023-03-31'],
      dtype='object', length=288)

Below we renamed the `RegionID` columns to `ZIPCODE` and select the `ZIPCODE`, `2023-01-31`, `2023-02-28`, and `2023-03-31` columns.  Then, we use the `groupby()` and `mean()` functions to calculate the means grouped by `ZIPCODE`.

Note that the `groupby()` function will indexed the results by the columns passed to the function.  In this case `ZIPCODE`.

In [16]:
house_df = (house_df
            .rename(columns={'RegionID':'ZIPCODE'})
            [['ZIPCODE','2023-01-31','2023-02-28','2023-03-31']]
            .groupby('ZIPCODE')
            .mean()
           )
house_df

2023-01-31     2023-02-28     2023-03-31
ZIPCODE                                             
58196    289183.650560  290142.126012  291579.076221
58197    452812.730506  453702.600673  456075.368414
58200    349605.853667  349729.762203  350587.034195
58201    402072.180437  401473.914859  401888.086022
58202    318573.721730  317472.656484  317132.928828
...                ...            ...            ...
777366   155659.981192  157189.842481  159339.001500
787970   525168.720819  524857.684731  527158.578055
787971   576787.298849  572369.045592  570487.717820
787972   418133.934425  418081.008836  419930.686376
808739   467690.950571  470487.848802  476631.332425

[27015 rows x 3 columns]

### Tax Data 
The tax data by zip code from the IRS was downloaded from https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-statistics-2020-zip-code-data-soi

Below we import the data and assign the `ZIPCODE` column as the index.

In [17]:
# tax_df = pd.read_csv('../data/20zpallnoagi.csv')
tax_df = pd.read_csv('https://raw.githubusercontent.com/emsanqui/Python_Training/master/data/20zpallnoagi.csv')
tax_df.tail()

STATEFIPS STATE  ZIPCODE  AGI_STUB       N1   MARS1   MARS2   MARS4   
27739         56    WY    83126         0    210.0    70.0   120.0     0.0  \
27740         56    WY    83127         0   1810.0   660.0  1020.0   100.0   
27741         56    WY    83128         0   1140.0   590.0   450.0    70.0   
27742         56    WY    83414         0    240.0    90.0   130.0     0.0   
27743         56    WY    99999         0  15470.0  7560.0  6540.0  1280.0   

           ELF  CPREP  ...  N85300  A85300  N11901   A11901   N11900   A11900   
27739    200.0    0.0  ...     0.0     0.0    40.0    236.0    140.0    442.0  \
27740   1670.0   40.0  ...    70.0   348.0   480.0   3071.0   1280.0   4708.0   
27741   1040.0   20.0  ...    70.0  3071.0   320.0   3970.0    780.0   4177.0   
27742    230.0    0.0  ...    40.0   608.0    60.0   1203.0    160.0   2515.0   
27743  14760.0  300.0  ...   510.0  3685.0  2810.0  18664.0  11660.0  53340.0   

        N11902   A11902  N12000   A12000  
27739    130.0    419.0     0.0      0.0  
27740   1240.0   4096.0    70.0    447.0  
27741    740.0   2302.0    70.0   1844.0  
27742    130.0    744.0    40.0   1725.0  
27743  11000.0  34850.0   630.0  15020.0  

[5 rows x 165 columns]

Since the tax data has 164 columns, we will select a few columns to work with.

In [18]:
tax_df.columns

Index(['STATEFIPS', 'STATE', 'ZIPCODE', 'AGI_STUB', 'N1', 'MARS1', 'MARS2',
       'MARS4', 'ELF', 'CPREP',
       ...
       'N85300', 'A85300', 'N11901', 'A11901', 'N11900', 'A11900', 'N11902',
       'A11902', 'N12000', 'A12000'],
      dtype='object', length=165)

Below we select the `ZIPCODE`, `A11902`, `N12000`, and `A12000` columns.  Then, we use the `groupby()` and `mean()` functions to calculate the means grouped by `ZIPCODE`.

In [19]:
tax_df = (tax_df[['ZIPCODE','A11902','N12000','A12000']]
          .groupby('ZIPCODE')
          .mean()
         )
tax_df

A11902        N12000        A12000
ZIPCODE                                          
0        7.038468e+06  79582.549020  1.549919e+06
1001     1.834100e+04    170.000000  6.760000e+02
1002     2.057400e+04    690.000000  6.983000e+03
1005     5.076000e+03     40.000000  1.000000e+02
1007     1.684800e+04    230.000000  1.268000e+03
...               ...           ...           ...
99827    2.387000e+03     40.000000  1.820000e+02
99833    3.040000e+03     80.000000  6.710000e+02
99835    9.920000e+03    170.000000  2.897000e+03
99901    1.414600e+04    160.000000  2.322000e+03
99999    1.046086e+05   1485.686275  4.731910e+04

[27644 rows x 3 columns]

## `join` Data Frame Method

Below we `inner` join the housing and tax data frames on `ZIPCODE`.  Other types of join types include `left`, `right`, `outer`, `inner`, `cross`, the default is `left`.  This works like a sql query join operator.

More information is available from the documentation https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html

In [20]:
house_df.join(tax_df, on='ZIPCODE', how='inner').tail()

2023-01-31     2023-02-28     2023-03-31         A11902   
ZIPCODE                                                               
99760    537968.316022  536433.957336  537070.542746     760.000000  \
99780    434185.921844  432841.520544  433126.001475    1744.000000   
99827    517890.313382  516713.553426  517114.077503    2387.000000   
99901    439281.300219  434493.659764  431751.812127   14146.000000   
99999    439127.352204  435375.871385  433760.080742  104608.627451   

              N12000        A12000  
ZIPCODE                             
99760       0.000000      0.000000  
99780       0.000000      0.000000  
99827      40.000000    182.000000  
99901     160.000000   2322.000000  
99999    1485.686275  47319.098039

## `concat` Pandas Function

More information is available from the documentation https://pandas.pydata.org/docs/reference/api/pandas.concat.html

The `concat` function does not use the data frame index.  Instead, we must use columns.  Below we concatenate the columns (`axis=1`) for the `house_df` and `tax_df` data frames using the `ZIPCODE` column.  Like `join` we can specify different types of joins.

In [21]:
pd.concat([house_df.reset_index(), tax_df.reset_index()], keys='ZIPCODE', join='inner', axis=1).tail()

Z                                                    I            
      ZIPCODE     2023-01-31     2023-02-28     2023-03-31 ZIPCODE   A11902   
27010  777366  155659.981192  157189.842481  159339.001500   97601  20285.0  \
27011  787970  525168.720819  524857.684731  527158.578055   97603  28046.0   
27012  787971  576787.298849  572369.045592  570487.717820   97621    228.0   
27013  787972  418133.934425  418081.008836  419930.686376   97623   2044.0   
27014  808739  467690.950571  470487.848802  476631.332425   97624   2650.0   

                      
      N12000  A12000  
27010  260.0  1875.0  
27011  310.0  1307.0  
27012    0.0     0.0  
27013    0.0     0.0  
27014   20.0    60.0

## `merge` Data Frame Method

In [22]:
house_df.reset_index().merge(right=tax_df.reset_index(), how='inner', on='ZIPCODE').tail()

ZIPCODE     2023-01-31     2023-02-28     2023-03-31         A11902   
6810    99760  537968.316022  536433.957336  537070.542746     760.000000  \
6811    99780  434185.921844  432841.520544  433126.001475    1744.000000   
6812    99827  517890.313382  516713.553426  517114.077503    2387.000000   
6813    99901  439281.300219  434493.659764  431751.812127   14146.000000   
6814    99999  439127.352204  435375.871385  433760.080742  104608.627451   

           N12000        A12000  
6810     0.000000      0.000000  
6811     0.000000      0.000000  
6812    40.000000    182.000000  
6813   160.000000   2322.000000  
6814  1485.686275  47319.098039